pkg/http/handlers/base.go

```diff
@@ -34,48 +36,107 @@ type BaseHandler interface {
+// `maxBodyBytes` is the maximal request body size, < 0 means the default Megabyte. Using 0 will disable the limit.
 // `componentName` is used for tracing to identify to which
 // component this handler belongs to.
-func NewBaseHandler(componentName string, maxBodyBytes int64, debug bool) BaseHandler {
+func NewBaseHandler(componentName string, maxBodyBytes int64, debug bool) *Handler {
 	if maxBodyBytes < 0 {
 		maxBodyBytes = Megabyte
 	}
-	return &baseHandler{
+	return &Handler{
 		Tracer:       tracing.NewTracer("handlers", componentName),
-		maxBodyBytes: maxBodyBytes,
-		debug:        debug,
+		MaxBodyBytes: maxBodyBytes,
+		Debug:        debug,
 	}
 }
 
 // NewBasehandlerWithTracer create a new base HTTP handler, like NewBaseHandler, but allows
 // the caller to configure the Tracer implementation independently.
-func NewBaseHandlerWithTracer(tracer tracing.Tracer, maxBodyBytes int64, debug bool) BaseHandler {
+//
+// Deprecated: you can now configure/override the default Tracer using
+//
+//    h := NewBaseHandler(componentName, maxBodyBytes, debug)
+//    h.Tracer = tracing.NewTracer("handlers", componentName)
+func NewBaseHandlerWithTracer(tracer tracing.Tracer, maxBodyBytes int64, debug bool) *Handler {
 	if maxBodyBytes < 0 {
 		maxBodyBytes = Megabyte
 	}
-	return &baseHandler{
+	return &Handler{
 		Tracer:       tracer,
-		maxBodyBytes: maxBodyBytes,
-		debug:        debug,
+		MaxBodyBytes: maxBodyBytes,
+		ErrorParser:  DefaultErrorParser,
+		Debug:        debug,
 	}
 }
 
-type baseHandler struct {
+// Handler is the default implementation of BaseHandler and is suitable for use in
+// most REST API implementations.
+type Handler struct {
 	tracing.Tracer
-	maxBodyBytes int64
-	debug        bool
+	// ErrorParser is used to parse error objects into HTTP status codes and response bodies.
+	ErrorParser ErrorParser
+	// MaxBodyBytes is the maximal request body size, < 0 means the default Megabyte.
+	// Using 0 will disable the limit and allow parsing streams.
+	MaxBodyBytes int64
+	// Debug was used to enable/disable Debug mode, when enabled error messages will be included in responses.
+	Debug bool
 }
 
-func (h *baseHandler) Error(ctx context.Context, w http.ResponseWriter, err error) {
+func (h *Handler) Error(ctx context.Context, w http.ResponseWriter, err error) {
 	span, _ := h.StartSpan(ctx, "Error")
 	defer h.FinishSpan(span, nil)
 
-	if err == nil {
-		w.WriteHeader(http.StatusOK)
+	parser := h.ErrorParser
+	if parser == nil {
+		parser = DefaultErrorParser
+	}
+
+	status, resp := parser(ctx, err, h.Debug)
+	h.Write(ctx, w, status, resp)
+}
+
+func (h *Handler) Write(ctx context.Context, w http.ResponseWriter, status int, obj interface{}) {
+	span, _ := h.StartSpan(ctx, "Write")
+	var err error
+	defer func() {
+		h.FinishSpan(span, err)
+	}()
+
+	if obj == nil {
+		w.WriteHeader(status)
 		return
 	}
 
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(status)
+	if obj != nil {
```

Comment:
That's always true at this point, isn't it?

Reply:
only for the default error handler, also the Write method can be called directly

Reply:
Sure, but the opposite check is done a few lines up and that case returns early. So this point here is never reached when `obj == nil` I think.

Reply:
updated